src/dotnet/CarbonAware.LocationSources.Azure/src/AzureLocationSource.cs

```diff
@@ -77,27 +40,39 @@ private string ReadFromResource(string key)
         return readerMetaData.ReadToEnd();
     }
 
-    protected virtual Dictionary<string, AzureRegion> GetAzureRegions()
+    protected virtual Dictionary<string, NamedGeoposition>? LoadRegionsFromJson()
     {
-        var data = ReadFromResource("azure-regions.json");
-        List<AzureRegion> regionList = JsonConvert.DeserializeObject<List<AzureRegion>>(data);
-        foreach(AzureRegion regionInfo in regionList)
-        {
-            Console.Write("region");
-            Console.WriteLine(regionInfo);
-        }    
-        // Create mapping
-        var azureRegionsMapping = new Dictionary<string, AzureRegion>();
-        foreach (var region in regionList)
-        {
-            azureRegionsMapping[region.Name] = region;
+        var data = ReadFromResource("CarbonAware.LocationSources.Azure.azure-regions.json");
+        List<NamedGeoposition> regionList = JsonSerializer.Deserialize<List<NamedGeoposition>>(data, options) ?? new List<NamedGeoposition>();
+        Dictionary<string, NamedGeoposition> namedGeopositions = new Dictionary<String, NamedGeoposition>();
+        foreach(NamedGeoposition region in regionList) {
+            namedGeopositions.Add(region.RegionName, region);
         }
+        return namedGeopositions;
+    }
 
-        // Cache mapping if didn't exist
-        if(!azureRegions.Any()) {
-           azureRegions = azureRegionsMapping;
+    public Location ToGeopositionLocation(Location location)
+    {
+        switch (location.LocationType) {
+            case LocationType.Geoposition: {
+                return location;
+            }
+            case LocationType.CloudProvider: {
+                if(location.CloudProvider != CloudProvider.Azure) {
+                    throw new ArgumentException($"Incorrect Cloud provider region. Expected Azure but found '{ location.CloudProvider }'");
+                }
+                NamedGeoposition region = namedGeopositions[location.RegionName];
```

Comment:
We should keep the check that ensures that the region is location in the dictionary, and throws an error if now (and adding logging)

Reply:
We throw an exception in the consumer of this code (WattTimeDatasource)

Reply:
I'm not sure that's actually what we want. How would the watt time data source know that the region name it passed here isn't a valid region name? That's something we would want to catch here, log, and then throw because this is the class that has the actual access to the azure region data.